File: Java/binarySearch/Template3/SearchForARange.java
package binarySearch.Template3;
//Given an array of integers nums sorted
// in ascending order, find the starting and ending position of a given target value.
// Split it into two Template 1
public class SearchForARange {

	public int[] searchForRange (int[] nums, int target) {
		int[] result = {-1, -1};
        if (nums.length == 0) return result;
        //if (nums.length == 1) {return new result[]{0, 0};}
        int left = 0, right = nums.length - 1;
        int start = -1, end = -1;
        // apply Template 1 to find the LEFT/RIHGT MOST target position
        while (left < right) {
            int mid = (left + right) >>> 1; // mid always biased towards LEFT
            if (nums[mid] < target) left = mid + 1;
            else right = mid;
        }
        if (nums[left] != target) return result;
        else start = left;
        
        right = nums.length - 1;
        while (left < right) {
            int mid = ((left + right) >>> 1) +1;//!!!!! make the mid biased towards right otherwise the search range might not move
            if (nums[mid] > target) right = mid - 1;
            else left = mid;// here tells why the above mid should ADD + 1. here LEFT == MID, So give [1, 2], the MID without adding 1
                            // also 1 which equals LEFT
        }
        end = left;
        
        result[0] = start;
        result[1] = end;
        
        return result;
	}
}

File: Python/BinarySearch/search_for_a_range.py
# -*- coding: utf-8 -*-
# =============================================================================
# Given an array of integers nums sorted in ascending order, find the starting and ending position of a given target value.
# 
# Your algorithm's runtime complexity must be in the order of O(log n).
# 
# If the target is not found in the array, return [-1, -1].
# 
# Example 1:
# 
# Input: nums = [5,7,7,8,8,10], target = 8
# Output: [3,4]
# Example 2:
# 
# Input: nums = [5,7,7,8,8,10], target = 6
# Output: [-1,-1]
# =============================================================================

def searchRange(nums, target) -> list:
    if not nums:
        return [-1, -1]
    left, right = 0, len(nums) - 1
    # binary search the left-most index
    left_most = helper(nums, left, right, target)
    # target is not available in nums
    # in ascending order => left = mid + 1
    if left_most >= len(nums) or nums[left_most] != target:
        return [-1, -1]
    
    # binary search the right-most index
    right_most = helper1(nums, left, right, target)
    return [left_most, right_most]
            
def helper(nums, left, right, target):
    while left <= right:
        mid = left + (right - left) // 2
        if nums[mid] == target:
            # DONOT RETURN
            # move left-side
            right = mid - 1 # the return is LEFT instead of MID
        elif nums[mid] > target:
            right = mid - 1
        else:
            left = mid + 1
            
    return left

def helper1(nums, left, right, target):
    while left <= right:
        mid = left + (right - left) // 2
        if nums[mid] == target:
            # DONOT RETURN
            # move right-side
            left = mid + 1
        elif nums[mid] > target:
            right = mid - 1
        else:
            left = mid + 1
    return left - 1

if __name__ == '__main__':
    rst = searchRange([5,7,7,8,8,10], 8)
    for i in rst:
        print(i)